Fixes #162 and #160

hoshinotsuyoshi opened the pull request:
<del>
Make order-complete hook methods(`generate_order_commissions` and `send_notification_mails_to_vendors`) be called in `Spree::Order#finalize!` instead of using `after_transition`. fixes https://github.com/spree-contrib/spree_multi_vendor/issues/162 and https://github.com/spree-contrib/spree_multi_vendor/issues/160
</del>


-----


<del>

So far, even if you don't have problems in production env, but in development env(usually `config.cache_class = false`), you will run into problems such as https://github.com/spree-contrib/spree_multi_vendor/issues/162 or https://github.com/spree-contrib/spree_multi_vendor/issues/160 .

It is hard (or difiicult) to avoid re-registering after_transition.

For example, some workaround such as https://github.com/spree-contrib/spree_multi_vendor/issues/162#issuecomment-774948105 will be needed.

This patch approach is focusing on that the `finalize!` method is already set with `after_transition to: :complete` by spree/spree's core.

Calling spree-multi-vendor's hook methods in `finalize` (a stable method, I think) will be a simple solution.

</del>

--------

Fixes #160 and #162 .


## PR description

### 1.

1st commit <del> 47cedb5eeaf08addc17897f0952d7554dbae48e8 </del> https://github.com/spree-contrib/spree_multi_vendor/pull/172/commits/e1879a22c74fed8d079eb08d156cb532b91f7206 simply adds a spec(regression test).  This spec is related to https://github.com/spree-contrib/spree_multi_vendor/issues/162 . And this spec will be passed. 

### 2.


In spec/dummy/config/environments/test.rb, if you set `config.cache_classes = false`, the spec will be failed.


```diff
Dummy::Application.configure do
  # Settings specified here will take precedence over those in config/application.rb

  # The test environment is used exclusively to run your application's
  # test suite. You never need to work with it otherwise. Remember that
  # your test database is "scratch space" for the test suite and is wiped
  # and recreated between test runs. Don't rely on the data there!
-  config.cache_classes = true
+  config.cache_classes = false
```


And a spec https://github.com/spree-contrib/spree_multi_vendor/blob/bf37f0624cc1a5f7a3b7431136cd8416a54fed60/spec/models/spree/order_spec.rb#L24-L36 will be failed, too. (This spec is related to https://github.com/spree-contrib/spree_multi_vendor/issues/160 .)

### 4.

the 2nd commit <del> https://github.com/spree-contrib/spree_multi_vendor/pull/172/commits/ab0dc2503a9d90b38a03ad5e85a1fb594e5029cc </del> https://github.com/spree-contrib/spree_multi_vendor/pull/172/commits/a5333cfa0ac21010f1693f121a009c10f5d360bd will solve both.

hoshinotsuyoshi committed:
Add a test which ensures mails will be sent to vendors only once per vendor

damianlegawiec commented on app/models/spree_multi_vendor/spree/order_decorator.rb:
```diff
@@ -1,8 +1,12 @@
-    base.state_machine.after_transition to: :complete, do: :send_notification_mails_to_vendors
+  end
+
+  def finalize!
```
@hoshinotsuyoshi I think we can avoid this by changing

```ruby
Spree::Order.prepend SpreeMultiVendor::Spree::OrderDecorator
```

to 

```ruby
Spree::Order.prepend SpreeMultiVendor::Spree::OrderDecorator unless Spree::Order.included_modules.include?(SpreeMultiVendor::Spree::OrderDecorator)

hoshinotsuyoshi replied:
👋  Thank you for your reviewing. 

OK.  I added commit about this. 1e30b4275fef436f4be654272b686cc71a4beaa

I think such a method was recently added to the guide.  :+1: https://github.com/spree/spree/commit/9ff42905579e2fda1c760fca528aae3aa822e29a

damianlegawiec replied:
@hoshinotsuyoshi so with this approach we could skip the `finalize!` overwrite altogether

hoshinotsuyoshi replied:
Ah, I may have misunderstood your intentions. 🙏 

Let me check about `unless Spree::Order.included_modules.include?(SpreeMultiVendor::Spree::OrderDecorator)` behavior whether is enough or not.   

I'll tackle this  on Friday or Saturday.
Stay tune!

hoshinotsuyoshi replied:
I confirmed the behavior. It seems to be OK. 

rebased && squashed && force-pushed. 

https://github.com/spree-contrib/spree_multi_vendor/pull/172/commits/e1879a22c74fed8d079eb08d156cb532b91f7206 
https://github.com/spree-contrib/spree_multi_vendor/pull/172/commits/a5333cfa0ac21010f1693f121a009c10f5d360bd

And edited PR description. https://github.com/spree-contrib/spree_multi_vendor/pull/172#issue-586246938

Thank you! @damianlegawiec

hoshinotsuyoshi committed:
Prepend the decorator only when not yet prepended

Fixes https://github.com/spree-contrib/spree_multi_vendor/issues/160
Fixes https://github.com/spree-contrib/spree_multi_vendor/issues/162

With following the guide: https://github.com/spree/spree/commit/9ff42905579e2fda1c760fca528aae3aa822e29a
